Add implication function

kirschmet2 opened the pull request:
-add boolean implication function called "implies" to Bits.hs
-test implies function using property testing

kirschmet2 committed:
Add implication function

-add boolean implication function called "implies" to Bits.hs
-test implies function using property testing

kirschmet2 committed:
undid accidental change

fricklerhandwerk commented on src/Bits.hs:
```diff
@@ -31,3 +31,8 @@ nor x y = not (or x y)
 xnor :: Bit -> Bit -> Bit
 xnor x y = not (xor x y)
+
+implies :: Bit -> Bit -> Bit
+implies I I = I
+implies I O = O
+implies O _ = I
```
What do you think about making it an operator? We could then write more naturally:

```haskell
if a => b then ...
```

On the other hand, we could reserve the operator notation for the more real-world-looking implementation in terms of `NAND`, and then test against the sane-looking one that uses pattern matching.

kirschmet2 replied:
We didn't use operator notation for the other functions either, I think it makes sense to use either verbs or signs for all the functions.

fricklerhandwerk commented on src/Bits.hs:
```diff
@@ -31,3 +31,8 @@ nor x y = not (or x y)
 xnor :: Bit -> Bit -> Bit
 xnor x y = not (xor x y)
+
+implies :: Bit -> Bit -> Bit
+implies I I = I
+implies I O = O
+implies O _ = I
```
Since you already said in the test what it does, why not implement it exactly that way?
```suggestion
implies :: Bit -> Bit -> Bit
implies I O = O
implies _ _ = I
```

That of course makes the test trivial to the point of being nonsensical, but that only says something about how powerful Haskell is.

The test is still valuable though if we wanted to change implementation later, say to express everything in terms of `NAND`, like in a real computer.

fricklerhandwerk commented on test/Main.hs:
```diff
@@ -20,6 +20,11 @@ testAnd a b
 
+testImplies :: Bit -> Bit -> Expectation
+testImplies a b
+  | a == I && b == O = implies a b `shouldBe` O
+  | otherwise = implies a b `shouldBe` I
```
```suggestion
  | a == I = implies a b `shouldBe` b
  | otherwise = implies a b `shouldBe` I
```
Here is a way to leave more room for the invariant, i.e. be less explicit and more general.

kirschmet2 committed:
Shorten implication function

Make use of the fact that there is only one combination of inputs with a different output, so only two lines are needed.

Co-authored-by: Valentin Gagarin <valentin@fricklerhandwerk.de>

kirschmet2 committed:
Remove unnecessary condition

If `a == I` output should indeed be b.

Co-authored-by: Valentin Gagarin <valentin@fricklerhandwerk.de>
